torch/csrc/autograd/input_metadata.h

```diff
@@ -75,37 +78,83 @@ struct InputMetadata {
   }
 
   at::Tensor zeros_like() const {
-    return at::zeros(shape_, options_);
+    TORCH_CHECK(
+        !is_nested_tensor(),
+        "Zeros is not currently supported for nested tensors.")
+    return at::zeros(shape_as_dim_vector(), options_);
   }
 
   bool is_same_shape(const at::Tensor& grad) const {
-    TORCH_CHECK(!grad.is_nested(), "Nested grads are not currently supported.")
-    return grad.sizes().equals(shape());
+    TORCH_CHECK(
+        grad.is_nested() == is_nested_tensor(),
+        "Both grad and InputMetadata need to be either nested or non nested tensors.")
+    if (grad.is_nested()) {
+      return at::native::get_nested_size_tensor(grad).is_same_size(
+          shape_as_tensor());
+    }
+    return grad.sizes().equals(shape_as_dim_vector());
   }
   bool is_expandable_to_shape(const at::Tensor& grad) const {
-    // TODO: Currently NestedTensors are not expandable.
-    return grad.is_nested() ? false
-                            : at::is_expandable_to(shape(), grad.sizes());
+    // Currently NestedTensors are not expandable. If this support is added then
+    // updates to reduce_grad will be needed
+    TORCH_CHECK(
+        grad.is_nested() == is_nested_tensor(),
+        "Both grad and InputMetadata need to be either nested or non nested tensors.")
+    return grad.is_nested()
+        ? false
+        : at::is_expandable_to(shape_as_dim_vector(), grad.sizes());
+  }
+
+  at::Tensor reduce_grad(at::Tensor& grad) const {
+    // Currently reduce_grad is only called if is_expandable_to_shape returns
+    // true For nested tensors this always returns False, so this check
+    // shouldn't fail
+    TORCH_INTERNAL_ASSERT(!grad.is_nested() && !is_nested_tensor())
+    return at::sum_to(std::move(grad), shape_as_dim_vector());
   }
 
   std::stringstream incompatible_shape_error_message(
       const size_t index,
       const at::Tensor& grad) const {
     std::stringstream ss;
-    TORCH_CHECK(!grad.is_nested(), "Nested grads are not currently supported.")
     ss << "invalid gradient at index " << index << " - got ";
-    ss << grad.sizes();
+    if (grad.is_nested()) {
+      ss << at::native::get_nested_size_tensor(grad);
```

Comment:
Should we be using `at::_nested_tensor_size` now?

Reply:
My thought here is since Christian is hoping that eventually nested_tensors store their size information in something other than a tensor we should keep the dispatcher out of the loop still. I guess in that case though the refactor would still be changing `at::native::get_nested_size_tensor` -> `at::native::get_nested_size_SOMETHING` so essentially the same change we would make here. Also probably is slightly more efficient to not go through the dispatcher. 

What do you think? I can be swayed on this one

Reply:
> eventually nested_tensors store their size information in something other than a tensor we should keep the dispatcher out of the loop still

is the issue that the object returned by the function is not something supported by the dispatcher?

Reply:
Yeah exactly its something called an `EfficientSizeNode` https://github.com/pytorch/nestedtensor/blob/a97e1f6fc1650372bcb149d1fc243dde87514dcf/nestedtensor/csrc/nested_tensor_impl.h#L78

Declartion: https://github.com/pytorch/nestedtensor/blob/a97e1f6fc1650372bcb149d1fc243dde87514dcf/nestedtensor/csrc/storage/EfficientSizeNode.h#L53

Reply:
I see. `at::_nested_tensor_size` would still eventually need to return some kind of representation that the dispatcher can understand right? wouldn't it be better to just have that function be the only function that has to deal with the representation of nested tensor sizes underneath whatever it actually is, and it would be the only place in the code responsible for spitting out handling it and spitting out something reasonable?

Reply:
From what I can tell when you call tensor.size() (with no int argument) or tensor.shape this is calling a handwritten python binding to the c++ code -> https://github.com/pytorch/pytorch/blob/master/torch/csrc/Size.cpp. I ran tensor.size() through lldb and the dispatcher does not get called at all. 

So I just get the feeling that if/when the DS changes from being a tensor, the dispatcher is not going to play a role- this feeling is also piggy backed off the idea that christian really didn't want to have this function exposed to any users. The tensor impl in that sense is the bastion of what sizes look like for nested tensors.